src/libraries/Common/src/System/Net/Http/Http2/Hpack/HPackDecoder.cs

```diff
@@ -450,8 +447,9 @@ int Decode(ref byte[] dst)
             {
                 if (_state == State.HeaderName)
                 {
-                    _headerNameLength = Decode(ref _headerNameOctets);
-                    _headerName = _headerNameOctets;
+                    var headerNameLength = Decode(ref _headerNameOctets);
+                    _headerName = WebHeaderEncoding.GetString(_headerNameOctets, 0, headerNameLength);
```

Comment:
I don't know that this is the way we should do this. You're trading performance here: dynamic headers that get reused will be cheaper now, but non-dynamic headers, or ones that don't get reused, will become more expensive.

HttpClient keeps its own interned strings around (the KnownHeader + known values table), so for many common headers it will not allocate any new strings. This change would instead allocate a string for every new header it sees.

A better solution might be to move the `KnownHeader` table into shared code and have `HPackEncoder` directly map to it when possible or allocate new strings if not. This would depend heavily on if Kestrel could make use of such a table. With some small changes this could also clean up a layering concern @Tratcher had re: exposing static indexes as an optimization.

Reply:
Maybe it can be solved together with https://github.com/dotnet/aspnetcore/issues/17400 as a plugin byte decoder. We can use our Known header optimizations in Libraries and Kestrel would use their own.